precompiles/xtokens/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+		input.expect_arguments(2)?;
+		let to_address: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		// We use the MultiLocationWrapper, which we have instructed how to read
+		// In the end we are using the encoding
+		let destination: MultiLocation = input.read::<MultiLocationWrapper>()?.into();
+
+		// Bound check
+		input.expect_arguments(1)?;
+		let weight: u64 = input.read::<u64>()?;
+
+		let to_account = Runtime::AddressMapping::into_account_id(to_address);
+		// We convert the address into a currency id xtokens understands
+		let to_currency_id: <Runtime as orml_xtokens::Config>::CurrencyId =
+			Runtime::account_to_currency_id(to_account)
+				.ok_or(error("cannot convert into currency id"))?;
+
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+		let to_balance = amount
+			.try_into()
+			.map_err(|_| error("Amount is too large for provided balance type"))?;
+
+		let call = orml_xtokens::Call::<Runtime>::transfer(
+			to_currency_id,
+			to_balance,
+			Box::new(destination),
+			weight,
+		);
+
+		let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+			Some(origin).into(),
+			call,
+			gasometer.remaining_gas()?,
+		)?;
+
+		gasometer.record_cost(used_gas)?;
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: Default::default(),
+			logs: Default::default(),
+		})
+	}
+
+	fn transfer_multiasset(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+
+		// asset is defined as a multiLocation. For now we are assuming these are concrete
+		// fungible assets
+		let asset_multilocation: MultiLocation = input.read::<MultiLocationWrapper>()?.into();
+		// Bound check
+		input.expect_arguments(1)?;
+		let amount: U256 = input.read()?;
+
+		// read destination
+		let destination: MultiLocation = input.read::<MultiLocationWrapper>()?.into();
+
+		// Bound check
+		input.expect_arguments(1)?;
+		let weight: u64 = input.read::<u64>()?;
+
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+		let to_balance = amount
+			.try_into()
+			.map_err(|_| error("Amount is too large for provided balance type"))?;
+
+		let call = orml_xtokens::Call::<Runtime>::transfer_multiasset(
+			Box::new(MultiAsset {
+				id: AssetId::Concrete(asset_multilocation),
+				fun: Fungibility::Fungible(to_balance),
+			}),
+			Box::new(destination),
+			weight,
+		);
+
+		let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+			Some(origin).into(),
+			call,
+			gasometer.remaining_gas()?,
+		)?;
+
+		gasometer.record_cost(used_gas)?;
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: Default::default(),
+			logs: Default::default(),
+		})
+	}
+}
+
+// A wrapper to be able to implement here the evmData reader
+pub struct MultiLocationWrapper(MultiLocation);
```

Comment:
Yeah, exactly this. Does it make sense to do this for the other data types? Maybe they don't need to be represented directly in solidity?

Reply:
they dont. The only one that needs to be represented ultimately is "MultiLocation", which wraps all the other types. But I will explore if I can make it through EVMData